server/dbversion.js

```diff
@@ -0,0 +1,86 @@
+const path = require('path')
+const Umzug = require('umzug')
+const db = require('./models')
+const fs = require('fs')
+
+// Sequelize uses umzug as the engine for running migrations, and stored completed migrations in a table
+// This module queries that table to see what migrations have been run against the database, and checks
+// against what migrations are available to this code. If there is a mismatch the throwAnyErrors method
+// will raise an error.
+// If sequelize stops using umzug in the future, umzug can be used to run the migrations directly so
+// we can have complete control of that part as well. We can also do this if it becomes necessary to
+// store things such as the date a migration took place, or was rolled back.
+
+class DbVersion {
+  constructor (options) {
+    this.availableVersions = []
+    this.highestVersion = ''
+    this.currentDatabaseVersion = ''
+    if (typeof options !== 'object') {
+      options = {}
+    }
+    if (typeof options.umzug !== 'object') {
+      options.umzug = new Umzug({
+        storage: 'sequelize',
+        storageOptions: {
+          sequelize: db.sequelize
+        },
+        migrations: {
+          path: path.join(__dirname, 'migrations')
+        }
+      })
+    }
+    this.umzug = options.umzug
+  }
+
+  async refreshCurrentDatabaseVersion () {
+    const result = await this.getLatestFromDB()
+    this.currentDatabaseVersion = result.length > 0 ? result[0].name : ''
+  }
+
+  async getLatestFromDB () {
+    try {
+      return await this.umzug.storage.model.findAll({
+        limit: 1,
+        order: [['name', 'DESC']]
+      })
+    } catch (err) {
+      console.log(err)
+    }
+  }
+
+  async refreshAvailableVersions () {
+    const migrationPath = path.join(__dirname, 'migrations')
+
+    this.availableVersions = fs
+      .readdirSync(migrationPath)
+      .filter(file => path.extname(file) === '.js')
+      .sort((a, b) => -a.localeCompare(b))

```

Comment:
I think my previous comment on this was misunderstood. We do need to check whether the database is on a higher version than the code but we don't need to manually read the file system to work that out.

Umzug's `pending` function reads the file system for available migrations, then filters out any that have already been run on the database. It seems cleaner and less error-prone to use the Umzug function, rather than write our own implementation of the same functionality.

Reply:
I re-wrote the code to not check the file system, and just use the `pending` function, but it did not work. The umzug "pending" code merges file system migrations and migrations already run from the database table, which isn't what we want. So the umzug "pending" method can't be used, we have to go to the filesystem.

Reply:
`umzug.pending` doesn't merge file system and database migrations, it filters file system migrations based on the database. The source code for it ends with this:
```
return all.filter((migration) => executedFiles.indexOf(migration.file) === -1)
```
where `all` is the file system migrations and `executedFiles` is from the database

Reply:
I mis-remembered. It doesn't merge the database in, but it only returns a list of files that haven't been migrated. So if you need a list of all the files (which we do, to check if the current database version is known to us) there is no way to get it.

Reply:
Ah I get it now! Sorry, my mistake - my suggestion of using `pending` doesn't make sense at all for this.